packages/schematics/package-update/utility/npm.ts

```diff
@@ -0,0 +1,266 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import { JsonObject, logging } from '@angular-devkit/core';
+import {
+  Rule,
+  SchematicContext,
+  SchematicsException,
+  Tree,
+  chain,
+} from '@angular-devkit/schematics';
+import * as http from 'http';
+import { Observable } from 'rxjs/Observable';
+import { ReplaySubject } from 'rxjs/ReplaySubject';
+import { concat, ignoreElements, map, mergeMap } from 'rxjs/operators';
+import * as semver from 'semver';
+
+const semverIntersect = require('semver-intersect');
+
+const kPackageJsonDependencyFields = [
+  'dependencies',
+  'devDependencies',
+  'peerDependencies',
+  'optionalDependencies',
+];
+
+
+const npmPackageJsonCache = new Map<string, Observable<JsonObject>>();
+
+function _getVersionFromNpmPackage(json: JsonObject, version: string, loose: boolean): string {
+  const distTags = json['dist-tags'] as JsonObject;
+  if (distTags && distTags[version]) {
+    return (loose ? '^' : '') + distTags[version] as string;
+  } else {
+    if (!semver.validRange(version)) {
+      throw new SchematicsException(`Invalid range or version: "${version}".`);
+    }
+    if (semver.valid(version) && loose) {
+      version = '~' + version;
+    }
+
+    const packageVersions = Object.keys(json['versions'] as JsonObject);
+    const maybeMatch = semver.maxSatisfying(packageVersions, version);
+
+    if (!maybeMatch) {
+      throw new SchematicsException(
+        `Version "${version}" has no satisfying version for package ${json['name']}`,
+      );
+    }
+
+    const maybeOperator = version.match(/^[~^]/);
+    if (version == '*') {
+      return maybeMatch;
+    } else if (maybeOperator) {
+      return maybeOperator[0] + maybeMatch;
+    } else {
+      return (loose ? '~' : '') + maybeMatch;
+    }
+  }
+}
+
+/**
+ * Get the NPM repository's package.json for a package. This is p
+ * @param {string} packageName The package name to fetch.
+ * @param {LoggerApi} logger A logger instance to log debug information.
+ * @returns {Observable<JsonObject>} An observable that will put the pacakge.json content.
+ * @private
+ */
+function _getNpmPackageJson(
+  packageName: string,
+  logger: logging.LoggerApi,
+): Observable<JsonObject> {
+  const url = `http://registry.npmjs.org/${packageName.replace(/\//g, '%2F')}`;
```

Comment:
This won't work in environments with that need proxies for npm (like in China iirc). I don't know what the solution is for that but would like to bring it up.

Reply:
No solution for that. Use some other tooling for now.